File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-ird-5-4/NonDeliverableSettlement.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class NonDeliverableSettlement : ISerialized
    {
        public NonDeliverableSettlement(XmlNode xmlNode)
        {
            XmlNode referenceCurrencyNode = xmlNode.SelectSingleNode("referenceCurrency");
            
            if (referenceCurrencyNode != null)
            {
                if (referenceCurrencyNode.Attributes["href"] != null || referenceCurrencyNode.Attributes["id"] != null) 
                {
                    if (referenceCurrencyNode.Attributes["id"] != null) 
                    {
                        referenceCurrencyIDRef_ = referenceCurrencyNode.Attributes["id"].Value;
                        Currency ob = new Currency(referenceCurrencyNode);
                        IDManager.SetID(referenceCurrencyIDRef_, ob);
                    }
                    else if (referenceCurrencyNode.Attributes["href"] != null)
                    {
                        referenceCurrencyIDRef_ = referenceCurrencyNode.Attributes["href"].Value;
                    }
                    else
                    {
                        referenceCurrency_ = new Currency(referenceCurrencyNode);
                    }
                }
                else
                {
                    referenceCurrency_ = new Currency(referenceCurrencyNode);
                }
            }
            
        
            XmlNode fxFixingDateNode = xmlNode.SelectSingleNode("fxFixingDate");
            
            if (fxFixingDateNode != null)
            {
                if (fxFixingDateNode.Attributes["href"] != null || fxFixingDateNode.Attributes["id"] != null) 
                {
                    if (fxFixingDateNode.Attributes["id"] != null) 
                    {
                        fxFixingDateIDRef_ = fxFixingDateNode.Attributes["id"].Value;
                        FxFixingDate ob = new FxFixingDate(fxFixingDateNode);
                        IDManager.SetID(fxFixingDateIDRef_, ob);
                    }
                    else if (fxFixingDateNode.Attributes["href"] != null)
                    {
                        fxFixingDateIDRef_ = fxFixingDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        fxFixingDate_ = new FxFixingDate(fxFixingDateNode);
                    }
                }
                else
                {
                    fxFixingDate_ = new FxFixingDate(fxFixingDateNode);
                }
            }
            
        
            XmlNode fxFixingScheduleNode = xmlNode.SelectSingleNode("fxFixingSchedule");
            
            if (fxFixingScheduleNode != null)
            {
                if (fxFixingScheduleNode.Attributes["href"] != null || fxFixingScheduleNode.Attributes["id"] != null) 
                {
                    if (fxFixingScheduleNode.Attributes["id"] != null) 
                    {
                        fxFixingScheduleIDRef_ = fxFixingScheduleNode.Attributes["id"].Value;
                        AdjustableDates ob = new AdjustableDates(fxFixingScheduleNode);
                        IDManager.SetID(fxFixingScheduleIDRef_, ob);
                    }
                    else if (fxFixingScheduleNode.Attributes["href"] != null)
                    {
                        fxFixingScheduleIDRef_ = fxFixingScheduleNode.Attributes["href"].Value;
                    }
                    else
                    {
                        fxFixingSchedule_ = new AdjustableDates(fxFixingScheduleNode);
                    }
                }
                else
                {
                    fxFixingSchedule_ = new AdjustableDates(fxFixingScheduleNode);
                }
            }
            
        
            XmlNode settlementRateOptionNode = xmlNode.SelectSingleNode("settlementRateOption");
            
            if (settlementRateOptionNode != null)
            {
                if (settlementRateOptionNode.Attributes["href"] != null || settlementRateOptionNode.Attributes["id"] != null) 
                {
                    if (settlementRateOptionNode.Attributes["id"] != null) 
                    {
                        settlementRateOptionIDRef_ = settlementRateOptionNode.Attributes["id"].Value;
                        SettlementRateOption ob = new SettlementRateOption(settlementRateOptionNode);
                        IDManager.SetID(settlementRateOptionIDRef_, ob);
                    }
                    else if (settlementRateOptionNode.Attributes["href"] != null)
                    {
                        settlementRateOptionIDRef_ = settlementRateOptionNode.Attributes["href"].Value;
                    }
                    else
                    {
                        settlementRateOption_ = new SettlementRateOption(settlementRateOptionNode);
                    }
                }
                else
                {
                    settlementRateOption_ = new SettlementRateOption(settlementRateOptionNode);
                }
            }
            
        
            XmlNode priceSourceDisruptionNode = xmlNode.SelectSingleNode("priceSourceDisruption");
            
            if (priceSourceDisruptionNode != null)
            {
                if (priceSourceDisruptionNode.Attributes["href"] != null || priceSourceDisruptionNode.Attributes["id"] != null) 
                {
                    if (priceSourceDisruptionNode.Attributes["id"] != null) 
                    {
                        priceSourceDisruptionIDRef_ = priceSourceDisruptionNode.Attributes["id"].Value;
                        PriceSourceDisruption ob = new PriceSourceDisruption(priceSourceDisruptionNode);
                        IDManager.SetID(priceSourceDisruptionIDRef_, ob);
                    }
                    else if (priceSourceDisruptionNode.Attributes["href"] != null)
                    {
                        priceSourceDisruptionIDRef_ = priceSourceDisruptionNode.Attributes["href"].Value;
                    }
                    else
                    {
                        priceSourceDisruption_ = new PriceSourceDisruption(priceSourceDisruptionNode);
                    }
                }
                else
                {
                    priceSourceDisruption_ = new PriceSourceDisruption(priceSourceDisruptionNode);
                }
            }
            
        
        }
        
    
        #region ReferenceCurrency_
        private Currency referenceCurrency_;
        public Currency ReferenceCurrency_
        {
            get
            {
                if (this.referenceCurrency_ != null)
                {
                    return this.referenceCurrency_; 
                }
                else if (this.referenceCurrencyIDRef_ != null)
                {
                    referenceCurrency_ = IDManager.getID(referenceCurrencyIDRef_) as Currency;
                    return this.referenceCurrency_; 
                }
                else
                {
                      return this.referenceCurrency_; 
                }
            }
            set
            {
                if (this.referenceCurrency_ != value)
                {
                    this.referenceCurrency_ = value;
                }
            }
        }
        #endregion
        
        public string referenceCurrencyIDRef_ { get; set; }
        #region FxFixingDate_
        private FxFixingDate fxFixingDate_;
        public FxFixingDate FxFixingDate_
        {
            get
            {
                if (this.fxFixingDate_ != null)
                {
                    return this.fxFixingDate_; 
                }
                else if (this.fxFixingDateIDRef_ != null)
                {
                    fxFixingDate_ = IDManager.getID(fxFixingDateIDRef_) as FxFixingDate;
                    return this.fxFixingDate_; 
                }
                else
                {
                      return this.fxFixingDate_; 
                }
            }
            set
            {
                if (this.fxFixingDate_ != value)
                {
                    this.fxFixingDate_ = value;
                }
            }
        }
        #endregion
        
        public string fxFixingDateIDRef_ { get; set; }
        #region FxFixingSchedule_
        private AdjustableDates fxFixingSchedule_;
        public AdjustableDates FxFixingSchedule_
        {
            get
            {
                if (this.fxFixingSchedule_ != null)
                {
                    return this.fxFixingSchedule_; 
                }
                else if (this.fxFixingScheduleIDRef_ != null)
                {
                    fxFixingSchedule_ = IDManager.getID(fxFixingScheduleIDRef_) as AdjustableDates;
                    return this.fxFixingSchedule_; 
                }
                else
                {
                      return this.fxFixingSchedule_; 
                }
            }
            set
            {
                if (this.fxFixingSchedule_ != value)
                {
                    this.fxFixingSchedule_ = value;
                }
            }
        }
        #endregion
        
        public string fxFixingScheduleIDRef_ { get; set; }
        #region SettlementRateOption_
        private SettlementRateOption settlementRateOption_;
        public SettlementRateOption SettlementRateOption_
        {
            get
            {
                if (this.settlementRateOption_ != null)
                {
                    return this.settlementRateOption_; 
                }
                else if (this.settlementRateOptionIDRef_ != null)
                {
                    settlementRateOption_ = IDManager.getID(settlementRateOptionIDRef_) as SettlementRateOption;
                    return this.settlementRateOption_; 
                }
                else
                {
                      return this.settlementRateOption_; 
                }
            }
            set
            {
                if (this.settlementRateOption_ != value)
                {
                    this.settlementRateOption_ = value;
                }
            }
        }
        #endregion
        
        public string settlementRateOptionIDRef_ { get; set; }
        #region PriceSourceDisruption_
        private PriceSourceDisruption priceSourceDisruption_;
        public PriceSourceDisruption PriceSourceDisruption_
        {
            get
            {
                if (this.priceSourceDisruption_ != null)
                {
                    return this.priceSourceDisruption_; 
                }
                else if (this.priceSourceDisruptionIDRef_ != null)
                {
                    priceSourceDisruption_ = IDManager.getID(priceSourceDisruptionIDRef_) as PriceSourceDisruption;
                    return this.priceSourceDisruption_; 
                }
                else
                {
                      return this.priceSourceDisruption_; 
                }
            }
            set
            {
                if (this.priceSourceDisruption_ != value)
                {
                    this.priceSourceDisruption_ = value;
                }
            }
        }
        #endregion
        
        public string priceSourceDisruptionIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-ird-5-4/NonDeliverableSettlement.cpp
// NonDeliverableSettlement.cpp 
#include "NonDeliverableSettlement.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

NonDeliverableSettlement::NonDeliverableSettlement(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //referenceCurrencyNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* referenceCurrencyNode = xmlNode->FirstChildElement("referenceCurrency");

   if(referenceCurrencyNode){referenceCurrencyIsNull_ = false;}
   else{referenceCurrencyIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- referenceCurrencyNode , address : " << referenceCurrencyNode << std::endl;
   #endif
   if(referenceCurrencyNode)
   {
      if(referenceCurrencyNode->Attribute("href") || referenceCurrencyNode->Attribute("id"))
      {
          if(referenceCurrencyNode->Attribute("id"))
          {
             referenceCurrencyIDRef_ = referenceCurrencyNode->Attribute("id");
             referenceCurrency_ = boost::shared_ptr<Currency>(new Currency(referenceCurrencyNode));
             referenceCurrency_->setID(referenceCurrencyIDRef_);
             IDManager::instance().setID(referenceCurrencyIDRef_,referenceCurrency_);
          }
          else if(referenceCurrencyNode->Attribute("href")) { referenceCurrencyIDRef_ = referenceCurrencyNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { referenceCurrency_ = boost::shared_ptr<Currency>(new Currency(referenceCurrencyNode));}
   }

   //fxFixingDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* fxFixingDateNode = xmlNode->FirstChildElement("fxFixingDate");

   if(fxFixingDateNode){fxFixingDateIsNull_ = false;}
   else{fxFixingDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- fxFixingDateNode , address : " << fxFixingDateNode << std::endl;
   #endif
   if(fxFixingDateNode)
   {
      if(fxFixingDateNode->Attribute("href") || fxFixingDateNode->Attribute("id"))
      {
          if(fxFixingDateNode->Attribute("id"))
          {
             fxFixingDateIDRef_ = fxFixingDateNode->Attribute("id");
             fxFixingDate_ = boost::shared_ptr<FxFixingDate>(new FxFixingDate(fxFixingDateNode));
             fxFixingDate_->setID(fxFixingDateIDRef_);
             IDManager::instance().setID(fxFixingDateIDRef_,fxFixingDate_);
          }
          else if(fxFixingDateNode->Attribute("href")) { fxFixingDateIDRef_ = fxFixingDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { fxFixingDate_ = boost::shared_ptr<FxFixingDate>(new FxFixingDate(fxFixingDateNode));}
   }

   //fxFixingScheduleNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* fxFixingScheduleNode = xmlNode->FirstChildElement("fxFixingSchedule");

   if(fxFixingScheduleNode){fxFixingScheduleIsNull_ = false;}
   else{fxFixingScheduleIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- fxFixingScheduleNode , address : " << fxFixingScheduleNode << std::endl;
   #endif
   if(fxFixingScheduleNode)
   {
      if(fxFixingScheduleNode->Attribute("href") || fxFixingScheduleNode->Attribute("id"))
      {
          if(fxFixingScheduleNode->Attribute("id"))
          {
             fxFixingScheduleIDRef_ = fxFixingScheduleNode->Attribute("id");
             fxFixingSchedule_ = boost::shared_ptr<AdjustableDates>(new AdjustableDates(fxFixingScheduleNode));
             fxFixingSchedule_->setID(fxFixingScheduleIDRef_);
             IDManager::instance().setID(fxFixingScheduleIDRef_,fxFixingSchedule_);
          }
          else if(fxFixingScheduleNode->Attribute("href")) { fxFixingScheduleIDRef_ = fxFixingScheduleNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { fxFixingSchedule_ = boost::shared_ptr<AdjustableDates>(new AdjustableDates(fxFixingScheduleNode));}
   }

   //settlementRateOptionNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* settlementRateOptionNode = xmlNode->FirstChildElement("settlementRateOption");

   if(settlementRateOptionNode){settlementRateOptionIsNull_ = false;}
   else{settlementRateOptionIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- settlementRateOptionNode , address : " << settlementRateOptionNode << std::endl;
   #endif
   if(settlementRateOptionNode)
   {
      if(settlementRateOptionNode->Attribute("href") || settlementRateOptionNode->Attribute("id"))
      {
          if(settlementRateOptionNode->Attribute("id"))
          {
             settlementRateOptionIDRef_ = settlementRateOptionNode->Attribute("id");
             settlementRateOption_ = boost::shared_ptr<SettlementRateOption>(new SettlementRateOption(settlementRateOptionNode));
             settlementRateOption_->setID(settlementRateOptionIDRef_);
             IDManager::instance().setID(settlementRateOptionIDRef_,settlementRateOption_);
          }
          else if(settlementRateOptionNode->Attribute("href")) { settlementRateOptionIDRef_ = settlementRateOptionNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { settlementRateOption_ = boost::shared_ptr<SettlementRateOption>(new SettlementRateOption(settlementRateOptionNode));}
   }

   //priceSourceDisruptionNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* priceSourceDisruptionNode = xmlNode->FirstChildElement("priceSourceDisruption");

   if(priceSourceDisruptionNode){priceSourceDisruptionIsNull_ = false;}
   else{priceSourceDisruptionIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- priceSourceDisruptionNode , address : " << priceSourceDisruptionNode << std::endl;
   #endif
   if(priceSourceDisruptionNode)
   {
      if(priceSourceDisruptionNode->Attribute("href") || priceSourceDisruptionNode->Attribute("id"))
      {
          if(priceSourceDisruptionNode->Attribute("id"))
          {
             priceSourceDisruptionIDRef_ = priceSourceDisruptionNode->Attribute("id");
             priceSourceDisruption_ = boost::shared_ptr<PriceSourceDisruption>(new PriceSourceDisruption(priceSourceDisruptionNode));
             priceSourceDisruption_->setID(priceSourceDisruptionIDRef_);
             IDManager::instance().setID(priceSourceDisruptionIDRef_,priceSourceDisruption_);
          }
          else if(priceSourceDisruptionNode->Attribute("href")) { priceSourceDisruptionIDRef_ = priceSourceDisruptionNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { priceSourceDisruption_ = boost::shared_ptr<PriceSourceDisruption>(new PriceSourceDisruption(priceSourceDisruptionNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<Currency> NonDeliverableSettlement::getReferenceCurrency()
{
   if(!this->referenceCurrencyIsNull_){
        if(referenceCurrencyIDRef_ != ""){
             return boost::shared_static_cast<Currency>(IDManager::instance().getID(referenceCurrencyIDRef_));
        }else{
             return this->referenceCurrency_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Currency>();
   }
}
boost::shared_ptr<FxFixingDate> NonDeliverableSettlement::getFxFixingDate()
{
   if(!this->fxFixingDateIsNull_){
        if(fxFixingDateIDRef_ != ""){
             return boost::shared_static_cast<FxFixingDate>(IDManager::instance().getID(fxFixingDateIDRef_));
        }else{
             return this->fxFixingDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<FxFixingDate>();
   }
}
boost::shared_ptr<AdjustableDates> NonDeliverableSettlement::getFxFixingSchedule()
{
   if(!this->fxFixingScheduleIsNull_){
        if(fxFixingScheduleIDRef_ != ""){
             return boost::shared_static_cast<AdjustableDates>(IDManager::instance().getID(fxFixingScheduleIDRef_));
        }else{
             return this->fxFixingSchedule_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableDates>();
   }
}
boost::shared_ptr<SettlementRateOption> NonDeliverableSettlement::getSettlementRateOption()
{
   if(!this->settlementRateOptionIsNull_){
        if(settlementRateOptionIDRef_ != ""){
             return boost::shared_static_cast<SettlementRateOption>(IDManager::instance().getID(settlementRateOptionIDRef_));
        }else{
             return this->settlementRateOption_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<SettlementRateOption>();
   }
}
boost::shared_ptr<PriceSourceDisruption> NonDeliverableSettlement::getPriceSourceDisruption()
{
   if(!this->priceSourceDisruptionIsNull_){
        if(priceSourceDisruptionIDRef_ != ""){
             return boost::shared_static_cast<PriceSourceDisruption>(IDManager::instance().getID(priceSourceDisruptionIDRef_));
        }else{
             return this->priceSourceDisruption_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PriceSourceDisruption>();
   }
}
}

